### PR TITLE
(WIP) Fix #1229: add managed dependency version (29.0-jre) for Guava

### LINKED ISSUE #1229: Replace direct Guava usage by Stargate (use shaded everywhere applicable) to reduce Guava conflict w/ Cassandra backend
There was a recent issue with an old Guava version (19.0) getting included due to exclusions, transitive dependencies, causing some test setup failures. Let's add a global managed version to try to remove/reduce likelihood of such problem resurfacing.

Although the first idea was same as #814 it turned out more difficult to try to align Guava version across C* 3.11, 4.0 and DSE (impossible, basically) and it appears much better to use Datastax OSS driver shaded Guava uniformly.
This lets underlying persistence/storage to use Guava version they do without affecting Stargate.
This may mean we require newer shaded version (current one is 25), but for now the version we have works fine.

### PR DESCRIPTION
**Pull request by tatu-at-datastax**
**What this PR does**:

Adds managed dependency version `29.0-jre` for Guava, removes explicit version definitions from poms

**Which issue(s) this PR fixes**:

Fixes #1299

**Checklist**
- [n/a] Changes manually tested -- dependency upgrade tested via existing unit, integ, func tests
- [n/a] Automated Tests added/updated -- same as above
- [n/a] Documentation added/updated -- should not change visible behavior.
-

### CODE REVIEW

**Comment by tatu-at-datastax** [comment]
Ready for review (CI passed).

**Review by ivansenic — CHANGES_REQUESTED** [review]
It would really be great if we could fix now also the apparently wrong imports of the shadowed guava.. I think every non-persistence module should use guava directly.. If the dependency is not in the pom explicitly we should add it..

Futhermore, maybe we can quickly investigate how can we ban this in future. The shadowed guava should not be on the compile path for those modules.. I have seen other deps to be also available in the compile, but then at run-time not.. I am pritty sure I hit that with `MutableInt (Commons Lang 2.6 API)`, but when using same from `lang3` it works..

@tatu-at-datastax adding request changes, but we can also split in another PR if you like it better..

**Comment by ivansenic** [comment]
@tatu-at-datastax also see Dependabot alert about the Guava version..

**Comment by tatu-at-datastax** [comment]
@ivansenic Ah yes. I was curious about shadowed guava and +1 for removing that if possible. I think I'd like to do that in a separate issue/PR.

Also agree with explicit dependency: I'll see if I can find usages, add those explicitly.
Guava as a dependency can be pretty gnarly.

**Comment by tatu-at-datastax** [comment]
Oh and Dependabot. Yeah, maybe it makes sense to boost default version to even higher -- if I recall, that CVE is not often actually relevant, but all security tools flag it so it's just easier to upgrade. We should be able to do that.

**Review by ivansenic — APPROVED** [review]
Changing to approved as we agreed to move suggestions to other PRs...

**Review by dougwettlaufer — APPROVED** [review]

**Comment by tatu-at-datastax** [comment]
@ivansenic @dougwettlaufer I will do some of simple changes here still, moving back to WIP. Apologies for confusion. Will let you know once this is ready

File: config-store-yaml/pom.xml [hunk 1]
```diff
@@ -41,7 +35,7 @@
       <dependency>
         <groupId>com.fasterxml.jackson.dataformat</groupId>
         <artifactId>jackson-dataformat-yaml</artifactId>
-        <version>2.11.3</version>
+        <version>2.12.5</version>
```
Comment (tatu-at-datastax) [inline config-store-yaml/pom.xml:38]: In near future, I'll probably try to add `dependencyManagement` for Jackson (similar to Guava here); challenges are similar as Stargate can use newer/-est versions just fine but C*3.11/4.0/DSE depend on older versions (although I am working on trying to get Jackson update for Cassandra too, concurrently).

File: grpc/pom.xml [hunk 2]
```diff
@@ -52,6 +52,10 @@
       <version>${grpc.version}</version>
     </dependency>
     <!-- Other -->
+    <dependency>
```
Comment (tatu-at-datastax) [inline grpc/pom.xml:55]: Has direct usage, adding dependency.

File: health-checker/pom.xml [hunk 3]
```diff
@@ -101,11 +95,6 @@
       <groupId>io.dropwizard</groupId>
       <artifactId>dropwizard-logging</artifactId>
     </dependency>
-    <dependency>
```
Comment (tatu-at-datastax) [inline health-checker/pom.xml:104]: Removed minimal usage (see below), and hence dependency and exclusions as well.

File: persistence-test/pom.xml [hunk 4]
```diff
@@ -52,9 +52,12 @@
       <scope>compile</scope>
     </dependency>
     <dependency>
+      <!-- 09-Sep-2021, tatu: Cassandra 3.11/4.0 backends depend on Jackson 2.9;
```
Comment (tatu-at-datastax) [inline persistence-test/pom.xml:55]: Could probably have worked around this by forcing `jackson-core` and `jackson-databind` versions to 2.12.5, but due to compatibility concerns with backends decided to limit scope here, only boosting to 2.9.10 that C*/DSE use.

File: pom.xml [hunk 5]
```diff
@@ -38,6 +38,7 @@
     <commons-io.version>2.7</commons-io.version>
     <immutables.version>2.5.6</immutables.version>
     <grpc.version>1.35.0</grpc.version>
+    <guava.version>30.0-jre</guava.version>
```
Comment (tatu-at-datastax) [inline pom.xml:41]: As per Dependabot warning about Guava version <= 29.0

**Comment by tatu-at-datastax** [comment]
@ivansenic @dougwettlaufer I did bit more work as suggested, added notes on changes -- please let me know if there's anything else to address here. I will create a new issue for removing/reducing use of shaded guava dependency for follow-up work (and maybe one for Jackson deps too)

File: cql/pom.xml [hunk 6]
```diff
@@ -63,6 +63,10 @@
       <artifactId>caffeine</artifactId>
       <version>2.8.5</version>
     </dependency>
+    <dependency>
```
[the remark this answers is not in the document]
Comment (ivansenic) [reply cql/pom.xml:66]: OK wait, but are you sure there is no transitive deps here of guava, meaning that something else was using Guava but now we override the version? I was thinking to add dep only where shadowed Guava was used and we wanted to move away from it.. Let's sync on this quickly..

File: health-checker/src/main/java/io/stargate/health/CheckerResource.java [hunk 7]
```diff
@@ -40,7 +39,7 @@ public CheckerResource() {
     // for production use with dependency injection
   }
 
-  @VisibleForTesting
+  //  @VisibleForTesting // the only use of Guava from this module
```
Comment (ivansenic) [inline health-checker/src/main/java/io/stargate/health/CheckerResource.java:42]: Delete the line?

File: grpc/pom.xml (hunk 2 above)
Comment (ivansenic) [reply grpc/pom.xml:55]: Here is the same, some gRPC dep was introducing guava and we now overwrote the version most likely

**Review by ivansenic — CHANGES_REQUESTED** [review]
I have some worries about this.. Let's sync please..

File: health-checker/src/main/java/io/stargate/health/CheckerResource.java (hunk 7 above)
Comment (tatu-at-datastax) [reply health-checker/src/main/java/io/stargate/health/CheckerResource.java:42]: > Delete the line?

Sure, will do.

**Comment by tatu-at-datastax** [comment]
@ivansenic I added explicit Guava dependency in cases where module itself depends on guava (grepped for imports), removed if not. So did not leave dependency for transitive usage.

**Comment by ivansenic** [comment]
@tatu-at-datastax What's the state with this and what path are we taking at the end.. Maybe we can talk shortly today?

**Comment by tatu-at-datastax** [comment]
@ivansenic yes, was about to add a note. Let's sync up today.

My plan was to remove usage of un-shaded Guava, along with reducing usage in general where applicable.

File: cql/pom.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (tatu-at-datastax) [reply cql/pom.xml:66]: Changed inclusion to shaded version.

File: grpc/pom.xml (hunk 2 above)
Comment (tatu-at-datastax) [reply grpc/pom.xml:55]: Changed the approach to avoid this issue.

**Comment by tatu-at-datastax** [comment]
Closing this in favor of #1253 (although going through similar issues over there -- but finally making progress).

**Pull request closed without merging**